src/vcf/normalizer.cpp

```diff
@@ -0,0 +1,189 @@
+
+        // This index is necessary for getting the samples where the mutated allele is present
+        for (size_t i = 0; i < record.alternate_alleles.size(); i++) {
+            std::string alternate = record.alternate_alleles[i];
+            std::string reference = record.reference_allele;
+            size_t position = record.position;
+
+            size_t corrected_position;
+            std::string corrected_alternate;
+            std::string corrected_reference;
+
+            // assertions / preconditions
+            if (alternate.size() < 1) {
+                throw new NormalizationError{record.line, "Alternate should not be empty"};
+            }
+            if (reference.size() < 1) {
+                throw new NormalizationError{record.line, "Reference should not be empty"};
+            }
+            if (reference == alternate) {
+                throw new NormalizationError{record.line, "Reference and alternate should not be identical"};
+            }
+
+            // count leading matching bases using mismatch with forward iterators
+            // these will point to the first leading mismatch. These will be the "start" in the .assign [start, end)
+            std::pair<std::string::iterator, std::string::iterator> lead_mismatch_indices =
+                    util::mismatch(reference.begin(), reference.end(),
+                                   alternate.begin(), alternate.end());
+
+
+            // count trailing matching bases using mismatch with reverse iterators
+            // the result are reverse_iterators, which later must be made forward iterators
+            std::pair<std::string::reverse_iterator, std::string::reverse_iterator> trail_mismatch_reverse =
+                    util::mismatch(reference.rbegin(), std::string::reverse_iterator(lead_mismatch_indices.first),
+                                   alternate.rbegin(), std::string::reverse_iterator(lead_mismatch_indices.second));
+            // trail_match_* point to the first match of the trailing bases. These will be the "end" in the .assign
+            // [start, end) below. "advance" because trail_mismatch_reverse.* were pointing to the last MISmatch
+            std::string::iterator trail_match_ref = make_forward_and_advance(trail_mismatch_reverse.first);
+            std::string::iterator trail_match_alt = make_forward_and_advance(trail_mismatch_reverse.second);
+
+
+            // using the previous iterators, assign only [first_mismatch, last_mismatch+1)
+            corrected_reference.assign(lead_mismatch_indices.first, trail_match_ref);
+            corrected_alternate.assign(lead_mismatch_indices.second, trail_match_alt);
+            corrected_position = position + (lead_mismatch_indices.first - reference.begin());
+
+            records.emplace_back(record.line, record.chromosome, corrected_position,
+                                 corrected_reference, corrected_alternate);
+        }
+
+        return std::move(records);
```

Comment:
Looks like this is not necessary/recommended because Clang is raising a warning: `moving a local object in a return statement prevents copy elision`. More info in http://vmpstr.blogspot.co.uk/2015/12/redundant-stdmove.html